lua/mason-null-ls/api/command.lua

```diff
@@ -0,0 +1,153 @@
+						return source_mapping.null_ls_to_package[package_name] ~= nil
+					end, package_names)
+
+					if #package_names == 0 then
+						return nil
+					end
+
+					return a.promisify(vim.ui.select)(package_names, {
+						prompt = ('Please select which source you want to install for language %q:'):format(
+							source_name
+						),
+						format_item = function(package_name)
+							local source_name = source_mapping.null_ls_to_package[package_name]
+							if registry.is_installed(package_name) then
+								return ('%s (installed)'):format(source_name)
+							else
+								return source_name
+							end
+						end,
+					})
+				end)
+			end)
+			:map(function(package_name)
+				return { package = package_name, version = version }
+			end)
+			:if_not_present(function()
+				notify(('Could not find Null-ls Source %q.'):format(source_name), vim.log.levels.ERROR)
+			end)
+	end, user_args)
+end
+
+---@async
+local function parse_packages_from_heuristics()
+	local source_mapping = require('mason-null-ls.mappings.source')
+	local registry = require('mason-registry')
+
+	-- Prompt user which source they want to install (based on the current filetype)
+	local current_ft = vim.api.nvim_buf_get_option(vim.api.nvim_get_current_buf(), 'filetype')
+	local filetype_mapping = require('mason-null-ls.mappings.filetype')
+	return Optional.of_nilable(filetype_mapping[current_ft])
+		:map(function(source_names)
+			return a.promisify(vim.ui.select)(source_names, {
+				prompt = ('Please select which source you want to install for filetype %q:'):format(current_ft),
+				format_item = function(source_name)
+					if registry.is_installed(source_mapping.null_ls_to_package[source_name]) then
+						return ('%s (installed)'):format(source_name)
+					else
+						return source_name
+					end
+				end,
+			})
+		end)
+		:map(function(source_name)
+			local package_name = source_mapping.null_ls_to_package[source_name]
+			return { { package = package_name, version = nil } }
+		end)
+		:or_else_get(function()
+			notify(('No Null-ls sources found for filetype %q.'):format(current_ft), vim.log.levels.ERROR)
+			return {}
+		end)
+end
+
+local parse_packages_to_install = _.cond({
+	{ _.compose(_.gt(0), _.length), parse_packages_from_user_args },
+	{ _.compose(_.equals(0), _.length), parse_packages_from_heuristics },
+	{ _.T, _.always({}) },
+})
+
+local NullLsInstall = a.scope(function(sources)
+	local packages_to_install = parse_packages_to_install(sources)
+	if #packages_to_install > 0 then
+		local registry = require('mason-registry')
+		_.each(function(target)
+			local pkg = registry.get_package(target.package)
+			pkg:install({ version = target.version })
+		end, packages_to_install)
+		local ui = require('mason.ui')
+		ui.open()
+		ui.set_view('All')
+		vim.schedule(function()
+			ui.set_sticky_cursor('installing-section')
+		end)
+	end
+end)
+
+vim.api.nvim_create_user_command('NullLsInstall', function(opts)
```

Comment:
any reason to not have this in the plugin folder? these commands are not available unless the user calls `setup`, which some may not want to

Reply:
That's precisely why, currently the commands can be relocated into the plugin folder with no consequences but later on when the null-ls integration arrives (`automatic_installation` option) they must be moved back where they are currently since they can only be used after the logic inside `setup` is executed

Reply:
I think in general it's good to keep all initialization code behind a mandatory `.setup()` function. The Vim-native runtimepath/packpath loading mechanism can be pretty difficult for people to grok, and it can be difficult to make things load in the order you expect/want it to. This also gives people full control, allowing lazy loading the plugin and whatnot. Also, although completely negligible in terms of performance, `plugin/` files can't be optimized through measures like impatient.nvim.

Reply:
makes sense, resolving